### ISSUE TITLE
Pallene closures

### ISSUE THREAD

**Issue by hugomg** [pallene-lang/pallene#174]
Pallene currently does not allow nested function definitions. They would be nice to have, but we have to figure out the implementation details first.

**Comment by srijan-paul** [comment]
Hi, I have a few very barebone suggestions,
One approach *might* be to have a global linked list of `TValue`s, and each closure (function) is called with an extra argument, which will be it's first upvalue. This is kind of similar to Upvalues in Lua, but can get ugly I guess. Plus the overhead is not that great.

So here is what I came up with, It would require an extra layer of inference in the type checker to keep track of which values are captured and maintain some metadata about them, and I'm not very sure about the implementation details. 

But after the codegen, I think we can forego the runtime cost of "closing" upvalues by walking over a linked list like in the Lua interpreter.

Simply put, we can consider this structure: 

```lua
local function make_adder(x: integer): integer->integer
  return function (y: integer): integer
     return x + y
  end
end

export function add(x: integer, y: integer): integer
   local addX: integer->integer = make_adder(x)
   return addX(y)
end
```

Ignoring Pallene's specifics, it can be unrolled to something like this: 

```c

// type decl for the upvalue
// list of the closure.
typedef struct {
  int x;
} t_closure_upvals;

// type signature of the anonymous function
typedef int (*t_closure_fn)(int, t_closure_upvals);

// Similar to Lua's Function object, a wrapper
// around a function and a list of it's upvalues.
typedef struct {
  t_closure_fn fn;
  t_closure_upvals upvals;
} t_closure;

// the anonymous function returned by make_adder
int closure(int y, t_closure_upvals upvals) { return upvals.x + y; }

// local function make_adder
t_closure make_adder(int x) {
  t_closure ret = {closure, {x}};
  return ret;
}

// export function add
int add(int x, int y) {
  t_closure add_x = make_adder(x);
  return add_x.fn(y, add_x.upvals);
}

int main() {
  printf("%d\n", add(2, 3));
  return 0;
}
```

This would mean, unlike Lua one would have to differentiate between which is a closure and which is a plain function.

**Nested closures**.

To implement closures that capture values from their outer scope, we can make that function's `t_closure_upvals` store a it's parent's upvalue list.

So something like: 

```lua
local function make_adder_2(x: integer): (integer) -> integer -> integer
  return function (y: integer): integer -> integer
    return function (z: integer): integer
      return x + y + z
   end
  end
end

export function add(x: integer, y: integer, z: integer): integer
   local add_x = make_adder_2(x)
   local add_x_y = add_x(y)
   return add_x_y(z)
end
```

Can be compiled to:

```c
#include <stdio.h>

// closure_1 = outer closure (int->int->int)
// closure_2 = inner closure (int->int)

typedef struct tcl2 t_closure_2;
typedef struct tcl1 t_closure_1;

typedef struct {
  int x;
} t_cl1_upvals;

typedef struct {
  int y; // variable captured from local scope.
  t_cl1_upvals parent; // upvalues captured by it's enclosing closure.
} t_cl2_upvals;

typedef t_closure_2 (*t_cl1_fn)(int, t_cl1_upvals);
typedef int (*t_cl2_fn)(int, t_cl2_upvals);

struct tcl1 {
  t_cl1_fn fn;
  t_cl1_upvals upvals;
};

struct tcl2 {
  t_cl2_fn fn;
  t_cl2_upvals upvals;
};

int cl2_fn(int z, t_cl2_upvals upvals) {
  return z + upvals.y + upvals.parent.x;
}

// lambda returned by make_adder_2
t_closure_2 cl1_fn(int y, t_cl1_upvals upvals) {
  t_closure_2 ret = {cl2_fn, {y, upvals}};
  return ret;
}

t_closure_1 make_adder_2(int x) {
  t_closure_1 ret = {cl1_fn, {x}};
  return ret;
}

int add(int x, int y, int z) {
  t_closure_1 add_x = make_adder_2(x);
  t_closure_2 add_x_y = x_add.fn(x, x_add.upvals);
  return add_x_y.fn(z, x_y_add.upvals);
}

// test
int main() {
  printf("%d\n", add(2, 3, 5));
  return 0;
}
```

I think this is similar to how [Nim implements closures](https://nim-by-example.github.io/procvars/).

**Comment by hugomg** [comment]
Hi srijan, I suspect that it might be possible to implement closures using the traditional technique of a closure-conversion transformation and an assignment-conversion for the mutable variables. Have you ever heard of these? Don't worry if you haven't. I'm planning to write a longer answer later but it will help if I know how familiar you are with that terminology.

One thing that I have been thinking about is that a first step towards full closures could be implementing closures without nested functions. Currently, we implement the toplevel module-local variables in a weird and ad-hoc manner. But it would make sense to implement them using closures, just like regular Lua does. If we can come up with a good design for Pallene closures then we might be able to implement it for toplevel variables first, as a proof of concept. If it works then we can expand the feature to also handle nested functions.

**Comment by srijan-paul** [comment]
>  Have you ever heard of these?

I'm afraid not. Do they have to do with Lambda lifting?
I've only ever implement closures using upvalues before, so that's the only technique I'm familiar with.

**Comment by hugomg** [comment]
Yeah, those are the basics. The annoying thing is that the terminology can vary depending on what source you are reading...

Lambda lifting, as used in the Nim docs, refers to rewriting the program without nested functions, using explicit closures. However, there are other places (such as [wikipedia](https://en.wikipedia.org/wiki/Lambda_lifting)) that use "lambda lifting" for something else: a way to implement nested functions without closures. Basically, in that version of lambda lifting we rewrite the program so that every inner function receives additional arguments corresponding to the outer variables. For example:

```lua
-- Before
local function f(x)
    local function g(y)
        return x + y
    end
    return g(1)
end

-- After
local function g(y, x)
    return x + y
end
function f(x)
    return g(1, x)
end
```
But that's enough of a digression :) The important thing that all approaches have in common is that there is some step that rewrites the code so that the functions are not nested anymore.

> I've only ever implement closures using upvalues before,

And to make things more confusing, even in Lua there is more than one meaning of "upvalue". [In the context of Lua functions](http://www.tecgraf.puc-rio.br/~lhf/ftp/doc/jucs05.pdf), upvalues are a pointer that points to a Lua variable from an outer function, which may be stored either in the Lua stack or in  the Lua heap. This implementation was chosen because it works well with Lua's single-pass compiler. When the compiler is generating the code for a local variable, it doesn't know yet if there will be a nested function further down which will try to access that variable.

However, in the context of the Lua-C API, there's another kind of upvalue. When we create a C closure using [lua_pushcclosure](https://www.lua.org/manual/5.4/manual.html#lua_pushcclosure) we can associate a set of values with it and these values can be manipulated by the C function using lua_getupvalue and lua_setupvalue. One important difference compared to the upvalues for Lua closures is that these upvalues are copied "by value". If we change them with lua_setupvalue, the outer function won't see that. In fact, in the context of the C API we don't even have the context of an "outer function". This definition of "upvalue" is closer to the [old kind of upvalue](https://www.lua.org/manual/4.0/manual.html#upvalue) that existed in Lua 4.0.

This is where the "assignment conversion" comes in. If we want to mutate an outer variable, one way to do it is by rewriting the code to store that value in a "box". For example:

    -- Normal code
    function f()
        local x = 10
        function g()
            x = x + 1
        end
     end
     
     -- Using a "box" for the mutable variable
     function f()
         local x = { contents = 10 }
         function g()
             x.contents = x.contents + 1
         end
      end

In the second version, the inner function never assigns to the variables of the outer function. Instead, it mutates the contents of those variables. This is compatible with an implementation that copies the upvalues "by value". 

By the way, if you are familiar with Python then you might have already seen this trick before -- it was the only way to mutate outer variables before they introduced the `nonlocal` keyword.

------------------

Going back to Pallene, the central constraint that we have to work around is that from the point of view of the Lua interpreter, Pallene functions are C functions and therefore, our closures have to be C closures. That is, the CClosure type from lobject.h.

This means that if we want to mutate the outer variable, we'll need to use that trick with the `x.contents`. Currently, we're sort of doing that for the global variables. All our closures have exactly one upvalue, which is a record holding all the global state: the toplevel variables declared with a `local` keyword, as well as references to all the string constants, metatables, and the closures for the toplevel functions.

What we need to figure out is how we're going to extend these ideas to work with nested functions in general. There are several questions to answer. The first is if we're going to use flat closures or linked closures.

![linked-closures](https://user-images.githubusercontent.com/1452192/108609751-d6ca0680-73ae-11eb-83a6-568aa5e6fe14.png)

(This picture is from [Efficient and safe-for-space closure conversion](https://sci-hub.se/10.1145/2661103.2661106), by Shao & Appel). Linked closures is what Nim does. The catch with them is that they can leak memory if we are not careful. For example, if `u` is a big table, the closure for `i` might keep that from being garbage collected even if the closure for `h` is not being directly referenced anymore.

Alternatively, we can have flat closures where every closure has a direct reference for every variable that they need. This seems closer to the upvalues that we have for C functions in Lua. The downside is that if there are a lot of variables in the outer scope then the closures can get big. It's possible that we'll want to store the global variables in a shared struct, sort of like the "safely linked" example.

-------

Another thing that we will need to figure out is the calling convention for Pallene functions. Every Pallene function has two "entry points", the Lua entry point and the Pallene entry point. For example, the following Pallene function

```lua
export function f(x: integer, y: integer): integer
    return x + y
end
```

is compiled to these C functions:

```c
static int function_02_lua(lua_State *L);

static lua_Integer function_02(
    lua_State *L,
    Udata *G,
    StackValue *base,
    lua_Integer x1, /* x */
    lua_Integer x2  /* y */);
```

When Lua wants to call the Pallene function it uses the Lua entry point, with the Lua calling convention: push the closure plus all the arguments to the Lua stack and then invoke `lua_call`.

However, pushing and popping things from the stack is slow. To avoid doing that all the time, when a Pallene function calls another Pallene function from the same module we use the "internal" Pallene calling convention. The function call is compiled to just a regular C function call with the appropriate arguments. If we want to implement nested functions we'll need to figure out what we need to change in this calling convention.

A refresher on the current calling convention: The `G` parameter refers to the record with the global state and the `base` parameter is the base of the Lua stack. From the point of view of Lua, a bunch of Pallene functions calling each other is one single C function call. The `base` refers to the first Pallene function that we called and that slot in the stack refers to the closure for that function.

Currently, the calling convention assumes that all the functions in the same module have the same upvalues. The `base` pointer points to the closure for the function at the base of the Pallene call stack and all the Pallene functions that we call use that same base pointer and that same closure. This will obviously have to change when we introduce nested functions but I don't know exactly how yet. Any suggestions would be welcome :)

---------------

The last thing that I wanted to talk about are [optimizations to reduce the cost of closures](https://sci-hub.se/10.1145/2661103.2661106)). For example, if we have a nested function that we know all the places that it is called, it might  be possible to use the "wikipedia version" of lambda lifting to avoid having to create a closure for it. If all the values it needs are passed as function arguments, they'll be passed via machine registers instead of being stored on the heap.

Now, I certainly think that we should complete the design for closures and nested functions before we worry too much about optimizing them. However, I think it would be a good idea to design our implementation so that it is works well with current optimizations (the internal Pallene calling convention) as well as future optimizations (such as getting rid of closures, or merging identical closures).

**Comment by srijan-paul** [comment]
Thank you very much for the detailed explanation!
I'm only familiar with lambda lifting as I've seen it used in the jlisp compiler and I have used Lua's upvalue design for my language. The paper on the other hand is new to me, so I'll be reading up a bit about closure conversions to approach this with a more well formed opinion and update this thread.

> However, in the context of the Lua-C API, there's another kind of upvalue.

This refers the `TValue` dynamic array in the `CClosure` right?

>  The function call is compiled to just a regular C function call with the appropriate arguments. If we want to implement nested functions we'll need to figure out what we need to change in this calling convention.

I like the idea of flat closures with extra arguments, though it does indeed have the downside of bloating the size with long upvalue chains like you said, and implementing higher order closures returning other closures that capture values from their original call site might be tricky this way, right?.  Or maybe it's just because a state + function bundle representation is easier for me to reason about at the moment. 

Using the boxing technique and storing the upvalues in an `object` also sounds good for an initial approach, although to use that in C using structs (since a `TValue` map would remove all the benefits of the explicit type annotations), I presume we might need to incrementally infer and build up the "shape" of the upvalues while type checking the AST? I can imagine that part becoming slightly simpler with safe linking. (Admittedly this is harder to garbage collect).

I might be missing some parts here so I'll let you know once I read the refs you provided.

**Comment by hugomg** [comment]
> The paper on the other hand is new to me

The paper about the Lua implementation is certainly required reading :)
The other ones that are linked can be useful as inspiration but we probably won't be doing exactly the same thing that they are doing. For example, the Shao paper that I took the pictures from is basically about a clever algorithm to figure out how which variables can be shared in that "safely linked" representation. In our case I suspect that a simple implementation with flat closures might suffice at first, because closures in Pallene wouldn't  be as performance-critical as they are in a LISP.

> This refers the TValue dynamic array in the CClosure right?

Exactly. Regular C code using the C API would access those slots using lua_getupvalue but in Pallene we can access the TValues directly.

>  though it does indeed have the downside of bloating the size with long upvalue chains like you said,

I expect that for the sort of programs that people use Pallene for, the closures will only be capturing a handful of free variables. But if that turns out to be an actual problem we can always think of ways to fix it later.

> implementing higher order closures returning other closures that capture values from their original call site might be tricky this way, right?

```lua
function outer()
   local x
   function middle()
       function inner()
           print(x)
       end
   end
end
```

The trick is that the middle function also has to capture all the variables used by the inner functions. In this example, both `middle` and `inner` would have `x` in their closure.

> although to use that in C using structs (since a TValue map would remove all the benefits of the explicit type annotations)

If we have a group of variables that are always captured together, then in theory the compiler could bundle them together in a Pallene record. That would use a C `struct` for the non-GC variables.

That said, for an initial design I'm more worried about coming up with the simplest thing that works (whatever that might be).

>  I presume we might need to incrementally infer and build up the "shape" of the upvalues while type checking the AST?

I expect that we'll need to make some changes to the type checker but I think it will be easier to know what after we figure out how our closures will be represented at runtime and at what step we're going to implement the lambda-lifting stuff.

> I saw in an old issue regarding the transpiler that Pallene was about to be in Google summer of code once before, would it be possible for me to suggest a project supplementing Pallene as mentioned here

No need to be bold. I was in fact going to bring up that topic at some point :)

This year, LabLua is applying to participate in GSOC but at the moment we are still waiting to see if our application will be accepted. That's why we haven't advertised the 2021 ideas list on our website yet. If we are accepted (fingers crossed...) then there are various kinds of Pallene-related projects that we could propose when applying for project slots. I usually add an idea or two to the list but some of the best project ideas are inspired by students.

Something involving closures would be an exciting project to tackle. The difficulty would be on the higher end of the scale but should be doable by a motivated student that's already familiar with the Pallene. The main catch is that in order for a GSOC project to go smoothly it's important to have a solid plan before the official coding period starts. Both the mentor and the student need to have a good idea of what code needs to be written as well as the expected timeline for writing said code.  In the case of Pallene closures there are still questions that we don't know the exact answer for and we'd have to figure those out ahead of time.

**Comment by hugomg** [comment]
Maybe the only change that we need to do for the calling convention is to pass all the variables from outer scope as extra arguments? For example, in a function like

```lua
function outer(a, b)
    function inner(c,d)
        return a+b+c+d
    end
end
```

The Pallene entry point for the inner function could look like this:

```c
static lua_Integer function_02(
    lua_State *L,
    StackValue *base,
    // Variables captured from outer scopes
    Udata *G,
    lua_Integer a,  
    lua_Integer b,
    // Usual parameters
    lua_Integer c,
    lua_Integer d);
```

The values of `G`, `a`, and `b` will be stored in the closure for the inner function. When the function is called from Lua, the Lua entry point can grab those values from the closure and pass them to the Pallene entry point.

The values of `G`, `a` and `b` are also known by the outer function, which needs to have them at hand in order to be able to construct the closure for the inner function. But if we call the inner function directly, without using the closure, then we can just pass them as arguments.

This of course assumes that the closures are "read only". Code using mutable values would need to be rewritten using an assignment conversion step, as we discussed before.

It does appear to work well with the optimizations of not creating the closure. If the inner function never "escapes", then we wouldn't need to create a closure for it.

**Comment by srijan-paul** [comment]
I went over some of the material you'd linked earlier. The Shao and Appel paper is interesting, but has some notation that I'm slowly getting used to. I think a classic version of lambda lift won't fit very well in C (as opposed to LISP dialects).

> This of course assumes that the closures are "read only".

I like this idea as it's more approachable to me, but having read-only closures like Python will then be something unique to Pallene, since Lua doesn't appear to have this restriction. If that's the direction Pallene wants to go then surely it's a good call!
Since closures aren't often used to model objects in Lua.

However, I was thinking about one possible way to implement closures that can encapsulate mutable table like behavior. 

> Pallene functions are C functions and therefore, our closures have to be C closures. That is, the CClosure type from lobject.h.

So I was wondering if there'll be any problems if we have 2 'versions' of these functions just like top level Pallene functions?

Pallene functions can be called from within two "contexts":
* From Lua's C API function `function_x_lua`.
* From within a Pallene function (`function_x`). 

In the second case, the function's type declaration looks something like so: 
```c
<return_type> function_x(lua_State* L,
        Udata* G,
        StackValue* base /* points to the slot in stack
			    slot the current CallInfo's
			    function lives */,
        <params>...     /* the parameters as defined in the
			   Pallene source */
); 
```
Typically, a Pallene function operates on the `<params>` and returns a C data type, which can then be used by the call site.

However, when a Pallene function is called from within Lua, we are required to follow a specific symbol type signature set specified by Lua.

```c
// Push the return values on the stack,
// and return the number if values returned.
int function_x_lua(lua_State* L);
```
Here, the parameters are assumed to be on the Lua stack, above the current base. They're extracted from there and the control again is deferred to `function_x`.

There is one constraint here that limits this to work for closures.
1. A closure has both code and data, the data needs to be passed down to the closure in some way, and would still need to maintain the `lua_State* -> int` type for the closures that are callable from Lua.
2. Closures can be created at runtime. They will share the function body but will have their own free variables that they'd want to close over.

Ideally, we would want to pass an extra parameter which would be the closure's environment that it has captured.

```c
// in a lua call
lua_Integer ret;
// something like this? 
int (*f(lua_State*, Udata*, StackValue, lua_Integer, TValue*)) 
	= cl->fn;
TValue* env = cl->env; // free vars captured by `cl`. 
ret = function_x(L, G, L->top, .. /* args */, env);
```
The disadvantage here is we're not leveraging the best of C by using the Lua C (`lua_Integer` etc) data types. But this way it might be easier to garbage collect, as p to structs, the type of which would need to be inferred.

For example, let's assume we want to pass a closure from Pallene to Lua.

```lua
export function make_adder(x: integer): integer -> integer
	return function (y: integer): integer
		return x + y
	end
end
```

In the type checking phase, we could mark nested functions as closures, that would then be accompanied by an environment. 

Some new IR instructions can then be used for mapping upvalue names to an index into the `TValue` list for the closure.

Assuming the above, we can aim for some C  code looking like so: 

```c
/* lambda in make_adder */ 
static lua_Integer lambda_01(lua_State *L,
	 Udata* G,
	 StackValue* base, 
	 lua_Integer x1 /* y */,
	 TValue* env) {
                lua_Integer x2 = ivalue(env[0]); /* upval x */
	        lua_Integer ret = intop(+, x1, x2);
	        return ret;
}

/* lambda in make_adder */
static int lambda_01_lua(lua_State* L) {
	CClosure* func = clCvalue(s2v(base));
	// Grab the environment (upval list) of the current
	// CClosure.
	TValue* env = func->upvalue;
	Udata *G = uvalue(env);
	
	int arg_count = lua_gettop(L);

	// Get the arguments from the stack
	lua_Integer x1 = ivalue(s2v(L->ci->func + 1));
	
	// Call the internal pallene version of the lambda and
	// pass the environment down.
	TValue ret = {{0}, 0};
	setivalue(&ret, lambda_01(L,G, L->top, x1));
	
	// Push the return value on the stack.
	setobj(L, s2v(L->top), &ret);
	L->top++;
	return 1;
}

/* Export */
/* make_adder */
static lua_Integer function_02(lua_State *L,
    Udata* G,
    StackValue* base,
    lua_Integer x1  /* x */) {
	 // push the upvalues onto the stack.
	 TValue x2 = {{0}, 0};
	 setivalue(&x2, x1);
	 set_obj(L, s2v(L->top), &x2);
	 L->top++;
	 
	 // make the closure.
	 lua_pushcclosure(L, lambda_01_lua, 1);
	 return 1;
}
```


Lastly, thank you for the heads up about gsoc, I'll apply this year and try! :crossed_fingers:

**Comment by hugomg** [comment]
> but having read-only closures like Python will then be something unique to Pallene, since Lua doesn't have this restriction

From the point of view of the Pallene programmer it would work like Lua; the inner function can mutate variables in the outer function. It's just that behind the scenes the compiler would transform it into an immutable variable pointing to a box with mutable contents.

I think the way that upvalues work in C closures pushes the design in this direction.

> There is one constraint here that limits this to work for closures
> Ideally, we would want to pass an extra parameter which would be the closure's environment that it has captured.

I think we won't need to change the Lua calling convention. The way that C closures work already takes care of the environment.

Remember that when we call a C function from Lua, we push into the stack the closure for that function followed by all the arguments. If we need to grab one of the upvalues, we can get it from the closure that's on the Lua stack.

In Pallene, you can see us doing that in our Lua entry points. We get the `G` record from one of the upvalues of the closure. What we'll need to do to implement nested functions is extend this so that the closures can also have other upvalues in addition to the `G` upvalue.

> Ideally, we would want to pass an extra parameter which would be the closure's environment that it has captured.

If we're using a "flat closures" model, then instead of passing one parameter for the environment, it might be more flexible to have one separate parameter for each variable that is captured.

If there are many variables, we still have the option to bundle them together in a record that is passed as a single parameter, like the "env" in this example.  But maybe we should think of that as an optimization instead of assuming that there is always going to be a single "environment" object.

> Assuming the above, we can aim for some C code looking like so:

That's actually very close to what I was thinking! What do you think of the following code? (I haven't tested it, but it should give the basic idea)

One difference in my version is that the "x" upvalue is passed "by value" because the adder never mutates the outer variable. If the lambda were to mutate the outer variable then the compiler would need to put the outer variable in a box, like we do with the `G`.

Another difference is that I'm setting two upvalues in the closure for the lambda, one for the G and one for the `x`. If we used a "linked" representation for the environment then it would make sense to have a single upvalue for the environment. But my hunch is that we probably want to go with a "flat closures" model to protect ourselves from space leaks.

Another thing that I would do differently is that if I had to pass an environment object down to the inner funtion then I would pass it as an `Udata*` instead of as a `TValue*`. If we need to do something to that environment, such as passing it somewhere or storing it in a child closure, then we'll need the userdata object itself. Basically what we are already doing with the `G`.

```c
/* Pallene entry point for the lambda */
static lua_Integer lambda_01(
        lua_State *L,
        StackValue* base,
        Udata* G,
        lua_Integer x1 /* x */,
        lua_Integer x2 /* y */)
{
    lua_Integer ret = intop(+, x1, x2);
    return ret;
}

/* Lua entry point for the lambda */
static int lambda_01_lua(lua_State* L)
{
    StackValue *base = L->ci->func;
    CClosure* func = clCvalue(s2v(base));

    // Grab the environment
    Udata *G       = uvalue(&func->upvalue[0]);
    lua_Integer x1 = ivalue(&func->upvalue[1]);

    // Get the arguments
    lua_Integer x2 = ivalue(s2v(base + 1));

    // Call the pallene entry point for the lambda,
    // passing the environment as additional arguments
    lua_Integer ret1 = lambda_01(L, L->top, G, x1, x2);

    // Push the return value to the stack
    setivalue(s2v(L->top), ret1);
    L->top++;
    return 1;
}

/* Pallene entry point for make_adder*/
static TValue function_02(
        lua_State *L,
        StackValue* base,
        Udata* G,
        lua_Integer x1  /* x */)
{
    TValue x2;

    // Create the C closure.
    // We'll use lua_pushcclosure. However, for the final implementation we
    // might want to use internal Lua functions such as luaF_newCclosure.
    // The kind of things that lua_pushcclosure is doing under the hood.
    setuvalue(L, s2v(L->top), G); L->top++;
    setivalue(s2v(L->top), x1); L->top++;
    lua_pushcclosure(L, lambda_01_lua, 2);
    // Get the closure
    L->top--;
    x2 = *s2v(L->top);

    return x2;
}
```

> In the type checking phase, we could mark nested functions as closures, that would then be accompanied by an environment.
> Some new IR instructions can then be used for mapping upvalue names to an index into the TValue list for the closure.

One thing that I'm wondering is if we should do these kind of lambda-lifting manipulations before converting the code to IR form or after. We'll need to think about this carefully. But I do think we're moving in the right direction.

> EDIT: I think the best way for me to further solidify this idea is to hand write some C code that does the same thing and test it. So I'll have to do that with whichever of the two ways we go.crossed_fingers

Absolutely! Experimenting with hand-written C code is the secret trick we always use when we want to add a new feature to Pallene :)

**Comment by srijan-paul** [comment]
That sounds ideal!
What do you think about doing the lambda lift in a separate pass over the AST after type checking?
Since that's when we have the most information about the inner lambdas (resolved parameter and return types etc).

For every nested call, we could add a lambda node to the `stats` list in the global scope. As a basic/rough idea:

1. In `parser.lua` we can add support for function expressions. So in the grammar, 
      - var ::= NAME | member_access | __lambda__

2. In `checker.lua`, the upvalues of a function can be identified. An `upvalues` field can be added to `ast.Exp.lambda` perhaps, to facilitate this. I think this might be the place to separate closures that mutate their upvalues (and hence need an 'environment') from flat closures that just get their upvalues handed down as parameters.
    - `Lambda = { "loc", "arg_decls", "upvalues", "body" } `.

3. The closure conversion pass (should this be a separate pass?), the lambdas can then be removed from the nested scope and added to the toplevel `stats`, replacing any old closure definitions with an ID perhaps? Alternatively, `to_ir.lua` could do the lifting whenever it comes across the lambda expression in an enclosing function's `stats` table.

4. In `to_ir.lua`, when we're emitting `CallStat` and `CallDyn`, we could add a case that handles calls to lambdas differently by passing along their extra arguments as needed.

**Comment by hugomg** [comment]
>  should this be a separate pass? Alternatively, to_ir.lua could do the lifting whenever it comes across the lambda expression in an enclosing function's stats table.

I feel that doing the lifting in an AST->IR pass (the `to_ir.lua`) sounds like the way to go.  The AST is supposed to map closely to actual syntax of Pallene, so if we wanted to make an AST->AST pass it would be awkward to represent the resulting closures as part of the AST. On the other end of the pipeline, the IR doesn't like nested functions so for an IR->IR pass it would be strange to represent the input IR before the lambda lifting.

Depending on the level of complexity that the lambda conversion ends up having, we can either do everything in the `to_ir.lua` or we can make an analysis pass first that adds annotations to the AST, similarly to what the `checker.lua` does.

> So in the grammar,  `var ::= NAME | member_access | lambda`

The `var` nonterminal refers to things that can appear in the left-hand-side of an assignment. I think in the case of lambdas what we'll want to do is add new cases under `exp` (for lambda expressions) and `stat` (for function definition statements).

A bonus stretch goal might be to update the grammar to make the toplevel of a Pallene module be a list of statements, just like in Lua.

> I think this might be the place to separate closures that mutate their upvalues (and hence need an 'environment') from flat closures that just get their upvalues handed down as parameters.

The "assignment conversion" part of the process, where we rewrite the mutable variables, does seem more flexible... I could see us doing it either in the AST or in the IR. Maybe we should ask @leokaplan to see if he has an opinion about this. He's currently working on implementing some optimizations at the IR level.

> when we're emitting CallStat and CallDyn, we could add a case that handles calls to lambdas differently by passing along their extra arguments as needed.

First of all, we'll need to add a `NewClosure` command. Currently, we're creating all the closures in the luaopen. One possible structure is to have this command receive an `f_id` (when same one we use for CallStatic) and a list of values corresponding to the upvalues.

    NewClosure   = {"loc", "f_id", "srcs"},

The `CallDyn` can stay unchanged. The `src_f` argument can be any Lua or Pallene closure and it shouldn't need to special case any of them.

For `CallStatic` we'll need to also pass the list of upvalues to it. One option is to add another field to the `CallStatic` command. Another option is to have the `srcs` list contain both the upvalues and the regular arguments.

**Comment by srijan-paul** [comment]
> I feel that doing the lifting in an AST->IR pass (the to_ir.lua) sounds like the way to go

Sure, we always have the option to move the assignment conversion out to `checker.lua` if that seems more conventional later.

> The var nonterminal refers to things that can appear in the left-hand-side of an assignment.

Whoops, true. I missed the `exp` production.

>Maybe we should ask @leokaplan to see if he has an opinion about this. He's currently working on implementing some optimizations at the IR level.

Surely. Given that, what direction do you feel is the best to go in, implementation wise? I was thinking of going topdown. Working our way down from the parser. For no particular reason other than It's easier for me personally, to reason about the process that way. Starting from the IR would not leave a lot of easy ways to test things, probably.

In the mean time, I'm focusing on the contingencies and playing around a bit with Pallene's headers and `LuaC_*`, to be able to draw parallels between the IR and C more clearly.

> First of all, we'll need to add a NewClosure command.

So this one generates the code for `luaF_newClosure`?

**Comment by hugomg** [comment]
> Surely. Given that, what direction do you feel is the best to go in, implementation wise? I was thinking of going topdown. Working our way down from the parser. 

I agree. This way it's easier to write the tests because you we can use Pallene code as the input for them.

That said, one thing that I like doing is adding an `error("not implemented")` stub to the later passes right away. Suppose that we change the definition of the `ast.Exp` to add another case. At the same moment I'd add an `error("not implemented")` case to any functions that are pattern matching over the `ast.Exp` tag. I'd rather see an error message saying that is is intentionally not implemented yet, instead of the default `tag_error` message.

> In the mean time, I'm focusing on the contingencies and playing around a bit with Pallene's headers and LuaC_*, to be able to draw parallels between the IR and C more clearly.

Great plan.

> So this one generates the code for luaF_newClosure

Exactly!

------------

By the way, I thought of a reason why we might want to do the assignment conversion stuff in the AST, before the conversion to IR. In Lua, the way that closures work cares about what scope your variable is declared. For example, in the following program the closures inside the loop care if their upvalue is declared in the block outside the loop or in the block inside the loop. The `fs` all return `4` at the end, while the `gs` print `1`, `2`, and `3`. However, in our IR we don't have any notion of block scope; each function in the IR has a single flat scope as if all the variables were declared at the start of the function.

```lua
local fs = {}
local gs = {}

do
    local i = 1
    while i <= 3 do
        fs[i] = function() return i end
        i = i + 1
    end
end

do
    local i = 1
    while i <= 3 do
        local j = i
        gs[j] = function() return j end
        i = i + 1
    end
end

print("fs:", fs[1](), fs[2](), fs[3]())
print("gs:", gs[1](), gs[2](), gs[3]())
```

This also gives me an idea of how we could hypothetically do that optimization of merging the boxes:

1. Two variables can only be stored in the same box if:
   1. They are declared in the same block scope.
   2. The closure(s) they appear in are garbage collected at the same time.
2. An upvalue that is mutable must be stored in a box.
3. An upvalue is immutable may be stored in a box, but doesn't have to.

We'd need an heuristic for rule 1.2 when the variables are used in more than one closure. One possibility is to say that the closures are collected together if they are all assigned to the same table. That should cover the common situation of returning a table of functions from module or using a table of functions for object oriented programming.

In rule 2, I'd guess that if we have two mutable upvalues with the same lifetime then it probably makes sense to put them in the same box.

For rule 3, there is a tradeoff: if we put an immutable variable in a box with the other variables then the function can receive one less argument. However, to use the variable we would need to look into the box, which might be more costly then just grabbing the variable from a function argument. We'd need to come up with a heuristic.

Nevertheless, we can probably ignore this optimization at first and only do it after we have a "minimum viable implementation" of closures and nested functions.

**Comment by srijan-paul** [comment]
> In Lua, the way that closures work cares about what scope your variable is declared.

Yeah, I was going to post a similar snippet. It's a bit convenient to work with nested function bodies instead of flat IR.

> This also gives me an idea of how we could hypothetically do that optimization of merging the boxes:

The rules make sense. We could use those as reference when optimizing boxes.
Would you say it would suffice to go with a full box even if a single free variable needs mutable capturing, as a first implementation?

---

So I was playing around a bit with Pallene snippets. I "hand compiled" and ran some closures in Pallene. One where the outer vars are mutated, and otherwise.

**Example 1**: Basic adder

```lua
export function make_adder(x: integer): integer -> integer
    return function (y: integer): integer
        return x + y
    end
end
```

This won't compile right now, so I compiled a boilerplate then added in some of the code by hand for experimenting. 
A paste of the C source is [here](https://pastebin.com/VKVbvAfn). The code was what we discussed earlier, and it seems to work with a very simple test script: 

![image](https://user-images.githubusercontent.com/50487716/109205053-1720ee80-77cc-11eb-9864-e1b0055e252a.png)
![image](https://user-images.githubusercontent.com/50487716/109205714-fa38eb00-77cc-11eb-9555-a3d7744dbf6f.png)


**Example 2:** Wrap a number in a box.

For this one, we have the following source:

```lua
export function wrap(n: integer): (integer -> (), () -> integer)
    local function set(x: integer) n = x end
    local function get(): integer return n end
    return set, get
end
```

I attempted modelling this one with a "box" like you mentioned earlier ([C source](https://pastebin.com/78Y4hjKs)). 
And tested with the following script:
![image](https://user-images.githubusercontent.com/50487716/109205865-30766a80-77cd-11eb-8d8a-1e58c25b4910.png)
![image](https://user-images.githubusercontent.com/50487716/109205886-379d7880-77cd-11eb-8c04-78584ca67867.png)

The only difference here is the userdata "box" around the value.

```c
/// closure get(): integer
static lua_Integer function_03(
    lua_State *L,
    Udata *G,
    StackValue *base,
    Udata *env
) {
    ptrdiff_t base_offset = savestack(L, base);
    return ivalue(&env->uv[0].uv);
}
```

Creating boxed closures gets a bit redundant with pushing `G` and `env` on and off for every closure, but I think this won't be an issue once we're using `luaF_newClosure` to cut dependence from the stack's behavior.

```c
/// wrap(n: integer): ((integer) -> (), () -> integer)
static TValue function_04(
    lua_State *L,
    Udata *G,
    StackValue *base,
    lua_Integer x1, /* x */
    TValue *ret2
) {
    ptrdiff_t base_offset = savestack(L, base);

    setuvalue(L, s2v(L->top), G); L->top++;
    // 1. Create the ENV
    lua_newuserdatauv(L, 0, 1);
    Udata *env = uvalue(s2v(L->top - 1));

    // 2. Store the upvalue in ENV
    setivalue(&(env->uv[0].uv), x1);

    // 3. Create a closure along with the upvalue
    lua_pushcclosure(L, function_03_lua, 2);
   
    // "return" this closure as the 2nd value
    *ret2 = *s2v(L->top - 1);

    // pop the value from the stack.
    lua_pop(L, 1);

    // Now we need to return the "set" closure.
    // 4. Push the G and ENV again
    setuvalue(L, s2v(L->top), G);   L->top++;
    setuvalue(L, s2v(L->top), env); L->top++;

    // 5. Create the set(x: integer) closure
    lua_pushcclosure(L, function_02_lua, 2);
    TValue ret = *s2v(L->top - 1);

    // pop that too
    lua_pop(L, 1);

    return ret;
}
```
So far these are just examples to demonstrate what compiled closures might look close to, and something for me to get familiar with atm. I presume the specifics of collecting CClosures and their upvalues remains to be figured out.

**Comment by hugomg** [comment]
>  Would you say it would suffice to go with a full box even if a single free variable needs mutable capturing, as a first implementation?

My vote would be to start with whatever is the simplest thing that could possibly work. My guess is that this would mean one upvalue per variable, with a separate box for each upvalue (irrespective or whether it's a mutable variable or not).

> I "hand compiled" and ran some closures in Pallene.

I like calling it "artisanal compiling" :)

> A paste of the C source is here.

Thanks, for uploading the full C code. I'll look at with carefully later. Please let me know if there if is something in particular that you want me to give an opinion about when I do that.

I also appreciate that you included the Lua scripts, although I'd prefer if they were included in a textual code block instead of as an image.

> Creating boxed closures gets a bit redundant with pushing G and env on and off for every closure

Separate upvalues for the G and for the `n` sounds about right if we're going with a flat-closures model.
One thing that might be illustrative would be to have an example where the outer function has more than one variable and each inner function uses different subset of the variables.

> I presume the specifics of collecting CClosures and their upvalues remains to be figured out.

Are you talking about garbage collection here?

**Comment by srijan-paul** [comment]
> My guess is that this would mean one upvalue per variable, with a separate box for each upvalue (irrespective or whether it's a mutable variable or not)

Sounds good. Is there any specific reason we're creating one box per variable and not boxing them together?

> although it might have been better to use textual code blocks instead of images

Ah, that is true.

```lua
-- test_adder.lua
local P = require "temp.adder"

for i = 1, 100 do
	local add_i = P.make_adder(i)
	for j = 1, 100 do
		assert(add_i(j) == i + j, "add failure @"..i..", "..j)
	end
end

print "Adder test Successful"
```

```lua
-- test_closure.lua
local P = require "temp.closure"

local set, get = P.wrap(0)

for i = 1, 100 do
	assert(get() == i - 1)
	set(i)
	assert(get() == i)
end

print "Boxed closure test successful"
```

>  an example where the outer function has more than one variable and each inner function uses different subset of the variables

Will do. 

> Are you talking about garbage collection here?

Yes, garbage collection.

**Comment by hugomg** [comment]
> Sounds good. Is there any specific reason we're creating one box per variable and not boxing them together?

I'd like to avoid memory leaks due to closures keeping references to upvalues that they don't actually need. This is what I mean when I say "safe for space". I remember this being a [big problem](https://mrale.ph/blog/2012/09/23/grokking-v8-closures-for-fun.html) in some other languages. In the linked post, check out the example with the `innerG` function which keeps the giant objects from being collected. Lua avoids this problem by storing each variable in a separate upvalue box instead of putting all of them in a big context object.

I understand the desire to use a single box for everything and indeed there are many common situations where doing so is perfectly safe for space (and possibly more efficient as well). However, in order to really be safe we'll need to analyze the code to ensure that the optimization is valid. If we want to start by implementing the simplest thing that could possibly work then I suspect that the way to go would be to use a separate box for each upvalue. That should be obviously safe for space without any analysis.

> Yes, garbage collection.

The golden rule for Pallene garbage collection is that every value that is garbage-collectable needs to be stored somewhere that the Lua GC can see it. Normally this means either the Lua stack or inside a Lua object, which is in turn stored in the Lua stack.

For our closures, if we implement everything using regular Lua data structures (C Closures and Userdatas) then in theory the garbage collection should do the right thing without any additional help from us. The main thing we'll need to be careful about is to avoid those memory leaks that I mentioned earlier.

**Comment by srijan-paul** [comment]
Ah I see. Just to be sure. We're relying on the GC's mark phase to mark the initial `CClosure` as a root, and then trace the upvalues incrementally from there, correct?

In short, we just have to careful to not allocate anything when the CClosure object isn't reachable as a root so as to not trigger a GC cycle?

Btw, Is there anything else you recommend I look into in the meantime, apart from the Shao and Appel paper?

**Comment by hugomg** [comment]
I got a chance to run the code examples that you uploaded earlier and they did work for me too. Exciting! 🚀🌜

Some comments:

* In the adder.c, there's a comment that says "Lua entry points" where you probably meant "pallene entry points". This part is correct in the closure.c
* Maybe we should order the function arguments so that the upvalues (n) appear before the regular variables (x). That might play nicer with variadic functions, if we ever add those to Pallene.
* In the closure.c, the code for creating the `set` and the `get` closures doesn't look the same. This might have to do with the hacky proof-of-concept implementation using existing  lua_api functions. One thing to consider would be to create a local variable to hold the `env`. That would allow us to reorder the operations. First, create the env and store it in a local variable. Then create the "get" closure and finally create the "set closure (in both cases using the logic that we are currently using for the "set").  Just make sure that you also save the value of the `env` to a slot in a Lua stack, instead of only putting it in a C local variable.
* In the closure.c we could consider using a Pallene record for the box instead of a hand-made userdata. Could be useful for artisanal compilation, if we want to use `--emit-c` to  generate the Pallene entry point for the inner lambda.

```
record IntBox
   n: integer
end
```

That said, a downside of using a Pallene record is that the current implementation of them always adds a metatable, to support run-time type checking. Those metatables would not necessary for boxes that are only used to implement our closures.

----------------------

> Ah I see. Just to be sure. We're relying on the GC's mark phase to mark the initial `CClosure` as a root, and then trace the upvalues incrementally from there, correct?

Precisely.

> In short, we just have to careful to not allocate anything when the CClosure object isn't reachable as a root so as to not trigger a GC cycle?

Exactly! The Pallene compiler normally takes care of this but when we get it wrong it can be a nasty bug where the GC deletes an object that we are still using... The gc.lua file has a nice comment discussing some of the things that we have to do so this part works properly.

When we implement closures I think that we won't need to mess with the GC-related parts of the compiler. That said, when doing "artisanal compilation" we might need to pay more attention to it. Just make sure that all the objects that you are working with are safely rooted on the Lua stack at any point where the Lua GC might run.

> Btw, Is there anything else you recommend I look into in the meantime, apart from the Shao and Appel paper? 

The paper about the Lua-C API might help explain some of the garbage collector things that we have been discussing. [Passing the language through the eye of a needle](https://dl.acm.org/doi/10.1145/1965724.1965739). In particular, the part where it talks about how the stack-based API helps keep all the Lua objects where the GC can see them.

[The implementation of Lua 5.0](http://www.tecgraf.puc-rio.br/~lhf/ftp/doc/jucs05.pdf), is a good reference for how Lua closures are implemented. They are not the same as the CClosures that we are using, but they also follow a "flat closures" model.

The part of the Shao and Appel paper that I think is most applicable for us is the discussion about flat closures vs linked closures. Other than that, a good chunk of the paper is talking about the algorithms they use to implement their "safely linked closures". That part probably doesn't apply to us because our requirement of reusing the existing CClosure datastructure pushes us strongly towards a simple flat-closures design.

**Comment by srijan-paul** [comment]
Thanks for the links, they were good reads. Especially the first one, never heard of it before. ^^

> existing CClosure datastructure pushes us strongly towards a simple flat-closures design.

To that end, what other major areas would you say remain to be discussed before one can start working on the implementation?

**Comment by srijan-paul** [comment]
Tried a closure with that captures only some of the variables from it's enclosing scope. I'm not sure if posting the C source will be useful since it's mostly the same.

> Or is it going to be a transformation step that does an AST->AST transformation?

I think one factor that affects the answer is how good this process fits a single pass.
Figuring out which variables from the outer scopes are captured appears to be comparatively straightforward (some slight modifications to the symbol table can do it, or keeping track of captured upvalues within a Lambda, like `lparser` does? ), so I guess the other part that dictates it is the actual assignment conversion. 

One way could be to do it in the `to_ir` pass, emitting two `SetUpVal` and `GetUpVal` IR instructions and do the conversions in the existing places that concern variable assignment and access (e.g `ToIR.exp_to_assignment`) as we come across them. 

An easier thing to do might be to test an implementation like Python where no mutation of captured variables is allowed, focusing only on upvalue access, and then introduce the mutation part once that is laid down? Until that is done, we could use a temporary "free variables captured from enclosing scopes cannot be mutated" error, or something similar.

An entire AST -> AST pass could possibly be avoided considering the above, but introducing the optimizations might change things, so I'm not totally sure about this part.

**Comment by hugomg** [comment]
> I'm not sure if posting the C source will be useful since it's mostly the same.

It's OK if you only post the things that need discussion. From my point of view, the most interesting thing about writing these examples by hand is having a good idea of what the result will look before you implement it.

> or keeping track of captured upvalues within a Lambda, like lparser does?

I feel like this part might belong in the type checker. We could a case for upvalues in `Checker.Name`, in addition to the cases for Local and Global that we already have.

> One way could be to do it in the to_ir pass, emitting two SetUpVal and GetUpVal IR instructions

Maybe we could avoid needing these instructions if the upvalues were treated like regular function parameters, which are treated as local variables by the IR (x1, x2, x3). After all, they're just that in the Pallene entry point.

I understand that we currently have special instructions for GetGlobal and SetGlobal but part of me wonders if we could get rid of those after we implement closures, turning the G into a regular upvalue.

> An easier thing to do might be to test an implementation like Python where no mutation of captured variables is allowed,

I love this idea. Yes, let's start by forbidding re-assignments to variables that are captured.

> Until that is done, we could use a temporary "free variables captured from enclosing scopes cannot be mutated" error, or something similar.

Additionally, the outer function is not allowed to mutate the variable.

> so I'm not totally sure about this part.

It might be easier to figure that out after we implement the closures with constant upvalues. Just need to ensure that it's part of the roadmap.

**Comment by srijan-paul** [comment]
> I feel like this part might belong in the type checker

Yeah, I meant resolving upvalues on-demand like lparser does, except in the checker :D 

> It might be easier to figure that out after we implement the closures with constant upvalues. Just need to ensure that it's part of the roadmap.

Agreed 👍

**Comment by hugomg** [comment]
I suppose we could do something like that. That said, the checker.lua is already one of the largest files in the compiler. By default I'd  prefer to put things in a separate pass where possible.

**Comment by srijan-paul** [comment]
Ah, fair.
So in case we decide upon an ast->ast pass after the checker, that would also require providing an extra step to `stop_after` in `driver.compile_internal`, right? (for testing, I imagine?). Any other parts of the codebase that this change would touch indirectly like this? And do we update all of the sites horizontally or add those in step by step (so the implement the basic AST pass first, then update the places where this new pass would then be used etc).

**Comment by hugomg** [comment]
> So in case we decide upon an ast->ast pass after the checker, that would also require providing an extra step to stop_after in driver.compile_internal, right?

Precisely.

> And do we update all of the sites horizontally or add those in step by step

The only important rule that we try to follow is that each PR is "self contained" and that all the tests pass before we merge it. But inside each PR the exact order doesn't matter as much. It's OK if the non-implemented stuff raises an `error("not implemented")` or if there are some test cases that are marked as `pending`.

For example, we could have a PR that updates the parser, another PR that implements closures for immutable variables, and another that implements mutable variables, and so on.

**Comment by srijan-paul** [comment]
Understood, thanks^^

**Comment by hugomg** [comment]
Good news w.r.t. GSOC. Lablua was accepted as an organization for 2021! We should probably start working a project proposal based on the [application template](http://www.lua.inf.puc-rio.br/gsoc/apply2021.html). If you are OK with that we can work on the proposal via email, to avoid cluttering the issue tracker.

**Comment by srijan-paul** [comment]
Awesome news!
I did drop an introduction in the mailing list a couple hours back.
I'll email about the proposal ^^

**Comment by hugomg** [comment]
Since this thread is getting long, we've decided to summarize the specification for this feature in a google doc:
https://docs.google.com/document/d/11eodot4ca2epSjRl4Xl0YLiWfcMCkNZ2FL00D9S7VN4

**Comment by hugomg** [comment]
Looks like I forgot to close this issue. Thanks again for implementing the closures, Srijan! :)
